torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,602 @@
-        r.add_(old_stps.at(i), (al.at(i) - be_i).item());
+      d = torch::mul(q, H_diag);
+      for (size_t i = 0; i < num_old; i++) {
+        auto be_i = old_dirs.at(i).dot(d) * ro.at(i);
+        d.add_(old_stps.at(i), val(al.at(i) - be_i));
       }
-      prev_flat_grad.copy_(flat_grad);
     }
 
-    /**
-     * compute step length
-     */
+    if (!prev_flat_grad.defined()) {
+      prev_flat_grad = flat_grad.clone(at::MemoryFormat::Contiguous);
+    } else {
+      prev_flat_grad.copy_(flat_grad);
+    }
+    prev_loss = loss;
 
+    // ############################################################
+    // # compute step length
+    // ############################################################
     // reset initial guess for step size
-    if (n_iter == 1) {
-      t = torch::min(ONE, ONE / abs_grad_sum) * options.learning_rate();
+    if (state.n_iter() == 1) {
+      t = std::min(1., 1. / val(flat_grad.abs().sum())) * lr;
     } else {
-      t = torch::tensor(options.learning_rate(), flat_grad.options());
+      t = lr;
     }
 
-    Tensor gtd = flat_grad.dot(d);
-    add_grad(t, d);
-    int64_t ls_func_evals = 0;
-    if (n_iter != options.max_iter()) {
-      // re-evaluate function only if not in last iteration
-      // the reason we do this: in a stochastic setting,
-      // no use to re-evaluate that function here
-      loss = closure();
-      flat_grad = gather_flat_grad();
-      abs_grad_sum = flat_grad.abs().sum();
-      ls_func_evals = 1;
-    }
+    // directional derivative
+    auto gtd = flat_grad.dot(d);  // g * d
 
-    current_evals += ls_func_evals;
+    // directional derivative is below tolerance
+    if (val(gtd) > -tolerance_change) break;
 
-    /**
-     * Check conditions
-     */
-
-    if (n_iter == options.max_iter()) {
-      break;
-    } else if (current_evals >= options.max_eval()) {
-      break;
-    } else if (abs_grad_sum.item<float>() <= options.tolerance_grad()) {
-      break;
-    } else if (gtd.item<float>() > -options.tolerance_grad()) {
-      break;
-    } else if (
-        d.mul(t).abs_().sum().item<float>() <= options.tolerance_change()) {
-      break;
-    } else if (
-        std::abs(loss.item<float>() - prev_loss.item<float>()) <
-        options.tolerance_change()) {
-      break;
+    // optional line search: user function
+    auto ls_func_evals = 0;
+    if (line_search_fn != c10::nullopt) {
+      TORCH_CHECK(*line_search_fn == "strong_wolfe", "only 'strong_wolfe' is supported");
+      auto x_init = _clone_param();
+      auto obj_func = [&](const std::vector<Tensor>& x, double t, const Tensor& d) { return _directional_evaluate(closure, x, t, d); };
+      auto strong_wolfe_res = _strong_wolfe(obj_func, x_init, t, d, loss, flat_grad, gtd);
+      loss = std::get<0>(strong_wolfe_res);
+      flat_grad = std::get<1>(strong_wolfe_res);
+      t = std::get<2>(strong_wolfe_res);
+      ls_func_evals = std::get<3>(strong_wolfe_res);
```

Comment:
We can write it as
```suggestion
      std::tie(loss, flat_grad, t, ls_func_evals) = _strong_wolfe(obj_func, x_init, t, d, loss, flat_grad, gtd);
```
to make it much shorter :D